runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10412,227 +10412,767 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+      0x67E12CDF, 887503681, 28629151, 923521, 29791, 961, 31, 1, 0, 0, 0, 0};
+   static uint32_t multiplierVectors_le16[16] = {0x94446F01, 0x94446F01, 0x94446F01, 0x94446F01,
+      1, 31, 961, 29791, 923521, 28629151, 887503681, 0x67E12CDF, 0, 0, 0, 0};
+   static uint32_t multiplierVectors_be32[12] = {923521, 923521, 923521, 923521,
+                                                29791, 961, 31, 1, 0, 0, 0, 0};
+   static uint32_t multiplierVectors_le32[12] = {923521, 923521, 923521, 923521,
+                                                1, 31, 961, 29791, 0, 0, 0, 0};
+   intptr_t multiplierVector;
+   switch (elementType)
+      {
+      case TR::Int8:
+         multiplierVector = (intptr_t) (isLE ?  multiplierVectors_le8 : multiplierVectors_be8);
+         break;
+      case TR::Int16:
+         multiplierVector = (intptr_t) (isLE ?  multiplierVectors_le16 : multiplierVectors_be16);
+         break;
+      case TR::Int32:
+         multiplierVector = (intptr_t) (isLE ?  multiplierVectors_le32 : multiplierVectors_be32);
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+   // point to the beginning of the array
+   loadAddressConstant(cg, false, node, multiplierVector, multiplierAddrReg);
+
+   // clear accumulator registers
+   switch (elementType)
+      {
+      case TR::Int8:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, third4Reg, third4Reg, third4Reg);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, fourth4Reg, fourth4Reg, fourth4Reg);
+         //fall through
+      case TR::Int16:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, high4Reg, high4Reg, high4Reg);
+         //fall through
+      case TR::Int32:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, low4Reg, low4Reg, low4Reg);
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+   // ready all zeros and ones registers
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, vconstant0Reg, vconstant0Reg, vconstant0Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vnor, node, vconstantNegReg, vconstant0Reg, vconstant0Reg);
+   // all four words in vunpackMaskReg a mask for the length of elementType
+   if (elementType != TR::Int32 && !isSigned)
+      {
+      generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::vsldoi, node, vunpackMaskReg, vconstant0Reg, vconstantNegReg, elementSize);
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::vspltw, node, vunpackMaskReg, vunpackMaskReg, 3);
+      }
+   // vend = end & (~0xf) = end of aligned data
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tempReg, 0xFFFFFFF0);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::AND, node, vendReg, endReg, tempReg);
+
+   // load the initial value
```

Comment:
there are pipeline nuances re short-latency vector instructions on POWER7 and POWER8.  most of permute instructions are short-latency instructions (with 2 cycle latency). if there are long-latency instructions (6-7 cycle latency) before them, it might cause a performance problem (result-bus collision leading to something so-called starvation of short-latency instructions). that could be the reason you observed inconsistent performance on POWER8. to avoid it, you have to insert regularly group-ending no-op (`ori r2, r2, 0 `  i believed). we might need to come back to this issue later. not now though.